components/fxa-client/ios/FxAClient/FxAccountState.swift

```diff
@@ -0,0 +1,102 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+/**
+ * States of the [FxaAccountManager].
+ */
+internal enum AccountState {
+    case start
+    case notAuthenticated
+    case authenticationProblem
+    case authenticatedNoProfile
+    case authenticatedWithProfile
+    case canAutoretryMigration
+}
+
+/**
+ * Base class for [FxaAccountManager] state machine events.
+ * Events aren't a simple enum class because we might want to pass data along with some of the events.
+ */
+internal enum Event {
+    case initialize
+    case accountNotFound
+    case accountRestored
+    case authenticateViaMigration(
+        sessionToken: String,
+        kSync: String,
+        kXCS: String
+    )
+    case authenticated(authData: FxaAuthData)
+    case authenticatedViaMigration
+    case authenticationError /* (error: AuthException) */
+    case recoveredFromAuthenticationProblem
+    case fetchProfile
+    case fetchedProfile
+    case failedToFetchProfile
+    case logout
+    case inFlightMigration
+    case retryMigrationLater
```

Comment:
It's not clear to me why `inFlightMigration` and `retryMigration` need to be two separate events here, what's the distinction?

Reply:
`inFlightMigration` is emitted on when the state machine initializes and notices that there is a migration in flight.
`retryMigrationLater` is emitted when trying to migrate and it fails.